### PR TITLE
Genetic music

### PR DESCRIPTION
**Pull request by JunShern**
New example demonstrating the use of genetic algorithms to generate musical content.

### CODE REVIEW

File: examples/genetic_music/sketch.js [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  }
+}
+
+function Earworm(indexNumber) {
+  this.id = indexNumber;
+  this.length = songLength;
+  this.notes = [];
+  // Visual properties
+  this.xpos = random(width);
+  this.ypos = random(height);
```
Comment (therewasaguy) [inline examples/genetic_music/sketch.js]: can we give meaning to the positioning? Perhaps tie it to the fitness score?

File: examples/genetic_music/sketch.js [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  this.ypos = constrain(this.ypos + random(-1, 1), 0, height);
+
+  push();
+  strokeWeight(1);
+  angleMode(DEGREES); // Change the mode to DEGREES
+  var angle = 360 / this.notes.length;
+  translate(this.xpos, this.ypos);
+  for (var i=0; i<this.notes.length; i++) {
+    var color = map(this.notes[i], minValidNote, maxValidNote, 280, 120) % 255;
+    var length = map(this.notes[i], minValidNote, maxValidNote, this.radius/2, this.radius);
```
Comment (therewasaguy) [inline examples/genetic_music/sketch.js]: nice mappings, these look great!

File: examples/genetic_music/sketch.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  this.xpos = constrain(this.xpos + random(-1, 1), 0, width);
+  this.ypos = constrain(this.ypos + random(-1, 1), 0, height);
+
+  push();
+  strokeWeight(1);
+  angleMode(DEGREES); // Change the mode to DEGREES
+  var angle = 360 / this.notes.length;
+  translate(this.xpos, this.ypos);
+  for (var i=0; i<this.notes.length; i++) {
+    var color = map(this.notes[i], minValidNote, maxValidNote, 280, 120) % 255;
```
Comment (therewasaguy) [inline examples/genetic_music/sketch.js]: if color corresponded to pitch class (0 - 11) that would help show patterns across octaves

File: examples/genetic_music/sketch.js [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  this.fitnessScore = 0;
+  // Key
+  for (var i=0; i<this.notes.length; i++) {
+    var keyClass = this.notes[i] % 12;
+    if (desiredKeyClasses.indexOf(keyClass) >= 0) {
+      this.fitnessScore = this.fitnessScore + 10;
+    }
+  }
+  // Prefer smaller intervals
+  for (var i=0; i<this.notes.length-1; i++) {
```
Comment (therewasaguy) [inline examples/genetic_music/sketch.js]: It seems like this has two goals:

1. Prefer melodies that are within a smaller range. Maybe 24 semitones (two octaves)
2. Within that range, prefer some intervals over others.

Smaller intervals are the most dissonant, so we might instead want to score based on the ratio of [consonant to dissonant intervals](https://www.howmusicreallyworks.com/Pages_Chapter_4/4_2.html#4.2.6).

File: examples/genetic_music/sketch.js [hunk 5]
```diff
@@ -0,0 +1,240 @@
+/**
+ *  Example: Genetic music
+ */
+
+var synth;
+var sloop;
+
+var validNotes = [...Array(128).keys()];
```
Comment (therewasaguy) [inline examples/genetic_music/sketch.js]: nice use of some ES6 features (spread operator and `.keys()`).

Some of these notes are really piercing, or inaudibly low. Maybe https://p5js.org/reference/#/p5/subset can help? I'd limit to those that appear on a piano or an organ (listed [here](http://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies))

File: examples/genetic_music/sketch.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  resetButton.mouseClicked(resetPopulation);
+  resetButton.position(10, 100);
+}
+
+function soundLoop(cycleStartTime) {
+  var duration = this.interval;
+  var velocity = 0.7;
+  var midiNote = population[clickedEarwormIndex].notes[notePlaybackIndex];
+  var noteFreq = midiToFreq(midiNote);
+  synth.play(noteFreq, velocity, 0, duration);
```
[the remark this answers is not in the document]
Comment (JunShern) [reply examples/genetic_music/sketch.js]: Oops! This was an accident, will fix this.

File: examples/genetic_music/sketch.js [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  // Move forward the index, and stop if we've reached the end
+  notePlaybackIndex++;
+  if (notePlaybackIndex >= population[clickedEarwormIndex].notes.length) {
+    this.stop();
```
[the remark this answers is not in the document]
Comment (JunShern) [reply examples/genetic_music/sketch.js]: Yup!

**Pull request merged**